reasoner/resolution/ResolverRegistry.java

```diff
@@ -18,51 +18,69 @@
 
+import grakn.common.collection.Pair;
 import grakn.common.concurrent.actor.Actor;
 import grakn.common.concurrent.actor.EventLoopGroup;
-import grakn.core.reasoner.resolution.framework.Answer;
+import grakn.core.logic.Rule;
+import grakn.core.logic.concludable.ConjunctionConcludable;
+import grakn.core.pattern.Conjunction;
+import grakn.core.pattern.equivalence.AlphaEquivalence;
+import grakn.core.reasoner.resolution.answer.MappingAggregator;
+import grakn.core.reasoner.resolution.framework.ResolutionAnswer;
 import grakn.core.reasoner.resolution.resolver.ConcludableResolver;
 import grakn.core.reasoner.resolution.resolver.RootResolver;
 import grakn.core.reasoner.resolution.resolver.RuleResolver;
+import graql.lang.pattern.variable.Reference;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 import java.util.HashMap;
-import java.util.List;
+import java.util.Map;
 import java.util.function.Consumer;
 
 public class ResolverRegistry {
 
     Logger LOG = LoggerFactory.getLogger(ResolverRegistry.class);
-    private final HashMap<Long, Actor<ConcludableResolver>> concludables;
-    private final HashMap<List<Long>, Actor<RuleResolver>> rules;
+    private final HashMap<ConjunctionConcludable<?, ?>, Actor<ConcludableResolver>> concludableActorsMap;
+    private final HashMap<Rule, Actor<RuleResolver>> rules;
     private final Actor<ResolutionRecorder> resolutionRecorder;
-    private final EventLoopGroup elg;
+    private EventLoopGroup elg;
 
     public ResolverRegistry(EventLoopGroup elg) {
         this.elg = elg;
-        concludables = new HashMap<>();
+        concludableActorsMap = new HashMap<>();
         rules = new HashMap<>();
         resolutionRecorder = Actor.create(elg, ResolutionRecorder::new);
     }
 
-    public Actor<ConcludableResolver> registerConcludable(Long pattern, List<List<Long>> rules, long traversalSize) {
-        LOG.debug("Register retrieval for concludable actor: '{}'", pattern);
-        return concludables.computeIfAbsent(pattern, (p) -> Actor.create(elg, self -> new ConcludableResolver(self, p, rules, traversalSize)));
+    public Pair<Actor<ConcludableResolver>, Map<Reference.Name, Reference.Name>> registerConcludable(ConjunctionConcludable<?, ?> concludable) {
+        LOG.debug("Register retrieval for concludable actor: '{}'", concludable.conjunction());
+        for (Map.Entry<ConjunctionConcludable<?, ?>, Actor<ConcludableResolver>> c: concludableActorsMap.entrySet()) {
+            AlphaEquivalence alphaEquality = c.getKey().alphaEquals(concludable);
+            if (alphaEquality.isValid()) {
+                return new Pair<>(c.getValue(), alphaEquality.asValid().namedVariableMapping());
+            }
+        }
+        Actor<ConcludableResolver> concludableActor = Actor.create(elg, self -> new ConcludableResolver(self, concludable));
+        concludableActorsMap.put(concludable, concludableActor);
+        return new Pair<>(concludableActor, MappingAggregator.identity(concludable));
```

Comment:
line reads odd, why are using a mapping aggregator to create a mapping hmmm maybe we can find a better naming after all

Reply:
Actually this should be ok, we just need to find a better place for the `identity` method, which returns an identity mapping. Already deliberated on where to put this, couldn't find somewhere better.

Reply:
If we still need to do a deep check after comparing with the hash, then the fastest we can do is to find the set of equal concludables and then run the deep check across those?

Reply:
yes, exactly, but better than looping. Was this response intended for another comment?

Reply:
Yes, above